Add SPI to clear data for WKWebsiteDataStore created with identifier

szewai opened the pull request:
#### a1dd9c5fa7efddf28bc7934f53cbaa8b6e49a27c
<pre>
Add SPI to clear data for WKWebsiteDataStore created with identifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=246282">https://bugs.webkit.org/show_bug.cgi?id=246282</a>
rdar://100920122

Reviewed by Geoffrey Garen.

Existing API removeDataOfTypes does not eliminate files and directories created for the WebsiteDataStore (i.e. it may
only truncate files), because it can be called while WebsiteDataStore is still in use. It&apos;s possible that the client
would not use the files any more and they should not be kept on disk, so adding an SPI for that.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/UUID.h:
* Source/WTF/wtf/cocoa/UUIDCocoa.mm: Added.
(WTF::UUID::operator NSUUID * const):
(WTF::UUID::UUID):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _fetchAllIdentifiers:]):
(+[WKWebsiteDataStore _removeDataStoreWithIdentifier:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration initWithIdentifier:]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::defaultWebsiteDataStoreRootDirectory):
(WebKit::WebsiteDataStore::fetchAllDataStoreIdentifiers):
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifier):
(WebKit::WebsiteDataStore::defaultWebsiteDataStoreDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):
(TestWebKitAPI::createWebsiteDataStoreAndPrepare):

Canonical link: <a href="https://commits.webkit.org/255520@main">https://commits.webkit.org/255520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f7fae711429c7f37c25a9df24a1856ccc23234c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92612 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102340 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1827 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30187 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85013 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98503 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98275 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1229 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79105 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28167 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83147 "Found 1 new API test failure: TestWebKitAPI.WKWebsiteDataStore.RemoveDataStoreWithIdentifier (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71258 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83978 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36595 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16778 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79021 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34383 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17958 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27402 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3821 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38253 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40567 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81641 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40162 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37116 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18467 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
```diff
@@ -81,6 +81,16 @@ - (void)webView:(WKWebView *)webView didReceiveAuthenticationChallenge:(NSURLAut
 }
 @end
 
+@interface WKWebsiteDataStoreMessageHandler : NSObject <WKScriptMessageHandler>
```
You could probably just use TestMessageHandler

achristensen07 commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -249,7 +249,39 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
+        return completionHandler("WebsiteDataStore with this identifier is in use"_s);
+
+    if (!FileSystem::deleteNonEmptyDirectory(defaultWebsiteDataStoreDirectory(identifier)))
+        return completionHandler("WebsiteDataStore with this identifier does not exist"_s);
```
It could exist but the delete operation could fail.

szewai replied:
Will change the error message to reflect that.

achristensen07 commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -249,7 +249,39 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
+{
+    bool isIdentifierInUse = false;
+    forEachWebsiteDataStore([&](auto& dataStore) {
+        if (auto storeIdentifier = dataStore.identifier(); storeIdentifier && *storeIdentifier == identifier)
```
I think std::optional's operator== should make it so you don't need to check for nullopt here.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -299,6 +299,35 @@ + (void)_deleteDefaultDataStoreForTesting
+        auto result = adoptNS([[NSMutableArray alloc] init]);
+        for (auto identifier : identifiers) {
+            auto uuid = adoptNS([[NSUUID alloc] initWithUUIDString:identifier.toString()]);
+            if (uuid)
```
if (auto uuid = ...)

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -299,6 +299,35 @@ + (void)_deleteDefaultDataStoreForTesting
+{
+    auto completionHandlerCopy = makeBlockPtr(completionHandler);
+    WebKit::WebsiteDataStore::fetchAllDataStoreIdentifiers([completionHandlerCopy](auto&& identifiers) {
+        auto result = adoptNS([[NSMutableArray alloc] init]);
```
initWithCapacity

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -299,6 +299,35 @@ + (void)_deleteDefaultDataStoreForTesting
+    WebKit::WebsiteDataStore::fetchAllDataStoreIdentifiers([completionHandlerCopy](auto&& identifiers) {
+        auto result = adoptNS([[NSMutableArray alloc] init]);
+        for (auto identifier : identifiers) {
+            auto uuid = adoptNS([[NSUUID alloc] initWithUUIDString:identifier.toString()]);
```
initWithUUIDBytes might skip the String roundtrip.
Also, WTF::UUID could be given a constructor that takes an NSUUID and a getter that returns a RetainPtr<NSUUID> to avoid duplicate code like this.

geoffreygaren commented on Source/WTF/wtf/UUID.cpp:
```diff
@@ -126,7 +126,12 @@ std::optional<UUID> UUID::parse(StringView value)
+std::optional<UUID> UUID::parse(StringView value)
+{
+    return UUID(convertStringToInt(value));
 }
```
This change doesn't seem right to me. The parse() function now unconditionally returns a UUID. That means the interface advertises an optional return value but never returns nullopt, and there's no explicit way to signal failure to parse.

I'd suggest retaining the old interface / behavior.

szewai replied:
Sure, will change.

geoffreygaren commented on Source/WTF/wtf/UUID.h:
```diff
@@ -54,6 +58,11 @@ WTF_MAKE_FAST_ALLOCATED;
     }
+    
+#ifdef __OBJC__
+    WTF_EXPORT_PRIVATE RetainPtr<NSUUID> createNSUUID() const;
```
Can this be an operator NSUUID * instead (like operator NSString * in String, which returns an autoreleased value)?

szewai replied:
Sure, will change.

geoffreygaren commented on Source/WTF/wtf/cocoa/UUIDCocoa.mm:
```diff
@@ -0,0 +1,46 @@
+// Use string instead of bytes to avoid consideration of endianess (NSUUID stores UUID as array of bytes,
+// and we store UUID as UInt128).
+UUID::UUID(NSUUID * nsUUID)
+    : UUID(convertStringToInt(String([nsUUID UUIDString])))
```
If you make the change above to revert the interface to parse(), then you can call parse() here and just unconditionally dereference the result, since we know an NSUUID is a valid UUID.

geoffreygaren replied:
Alternatively, if we believe there is some reason NSUUID might sometimes fail to convert to UUID, then we should use an explicit "from(NSUUID *)" interface that returns an optional<UUID>.

szewai replied:
I think NSUUID is a valid UUID so I will go with option 1.

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -299,6 +299,33 @@ + (void)_deleteDefaultDataStoreForTesting
+    WebKit::WebsiteDataStore::fetchAllDataStoreIdentifiers([completionHandlerCopy](auto&& identifiers) {
+        auto result = adoptNS([[NSMutableArray alloc] initWithCapacity:identifiers.size()]);
+        for (auto identifier : identifiers) {
+            if (auto uuid = identifier.createNSUUID())
```
When does create return nil? Why would skipping in that case be OK?

szewai replied:
This should not return nil, since fetchAllDataStoreIdentifiers returns valid UUID, so I will just remove this.

geoffreygaren commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -249,7 +249,42 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
+            isIdentifierInUse = true;
+    });
+    if (isIdentifierInUse)
+        return completionHandler("WebsiteDataStore with this identifier is in use"_s);
```
I think it would be better in this case just to go ahead with the delete operation. It's not in a programmer's control to ensure that the last reference to a reference counted object has been dropped. And the delete will do what the programmer intended, so long as the programmer does not do new things to trigger storage after the delete.

geoffreygaren replied:
(I'm expecting that removing a data store and then performing a storage operation on it would have the same effect as removing a data store and then invoking the dataStoreWithIdentifier: API with the same identifier and then performing a storage operation on the new data store -- namely, it would re-create that data store identifier on disk.)

szewai replied:
Sure, updated.

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -299,6 +299,33 @@ + (void)_deleteDefaultDataStoreForTesting
+    });
+}
+
++ (void)_deleteDataStoreWithIdentifier:(NSUUID *)identifier completionHandler:(void(^)(NSError* error))completionHandler
```
I would call this "_removeDataStoreWithIdentifier", since remove is the term we use for other data deletion in this API.

szewai replied:
Sure.

geoffreygaren reviewed (changes requested):
I think this is close but not quite right yet.

geoffreygaren commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -249,7 +249,34 @@ static bool experimentalFeatureEnabled(const String& key, bool defaultValue = fa
+    completionHandler(WTFMove(identifiers));
+}
+
+void WebsiteDataStore::removeDataStoreWithIdentifier(const UUID& identifier, CompletionHandler<void(const String&)>&& completionHandler)
```
Not for this patch, but: I wonder if we should do anything to post a message to the Network process and wait for an async reply, to make sure that any pending storage operations have been flushed before we attempt to remove the directory. The most basic solution would just post to the main thread of the Network process, to make sure to clear out any pending IPC. Even more complete would be to post a task to each storage type thread and wait for that set of tasks. It feels like a tiny bit of effort here may be justified, because I think a simple test where you just set a cookie and then delete a data store may fail in the current implementation (since the cookie set is an async IPC and the delete is a synchronous system call).

szewai replied:
> I think a simple test where you just set a cookie and then delete a data store may fail in the current implementation

Do you mean the directory might be failed to delete?

I guess it makes sense to ask child processes to clear data (or finish existing operations) before UI process deletes the directory when `WebsiteDataStore` with target identifier is alive. But even network process finishes task scheduled before `removeDataStoreWithIdentifier`, there seems no guarantee that network process will no longer write to that WebsiteDataStore's directories (e.g. ServiceWorker process updates LocalStorage).
Do you think we should forbid network process running storage operations on the session (WebsiteDataStore) after network process replies okay to `removeDataStoreWithIdentifier` message? We might achieve that by revoking sandbox extension to paths I guess.

geoffreygaren reviewed (approved):
r=me

szewai committed:
Add SPI to clear data for WKWebsiteDataStore created with identifier
https://bugs.webkit.org/show_bug.cgi?id=246282
rdar://100920122

Reviewed by Geoffrey Garen.

Existing API removeDataOfTypes does not eliminate files and directories created for the WebsiteDataStore (i.e. it may
only truncate files), because it can be called while WebsiteDataStore is still in use. It's possible that the client
would not use the files any more and they should not be kept on disk, so adding an SPI for that.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/UUID.h:
* Source/WTF/wtf/cocoa/UUIDCocoa.mm: Added.
(WTF::UUID::operator NSUUID * const):
(WTF::UUID::UUID):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::toImpl):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(+[WKWebsiteDataStore _fetchAllIdentifiers:]):
(+[WKWebsiteDataStore _removeDataStoreWithIdentifier:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.mm:
(-[_WKWebsiteDataStoreConfiguration initWithIdentifier:]):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::defaultWebsiteDataStoreRootDirectory):
(WebKit::WebsiteDataStore::fetchAllDataStoreIdentifiers):
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifier):
(WebKit::WebsiteDataStore::defaultWebsiteDataStoreDirectory):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp:
(WebKit::WebsiteDataStoreConfiguration::WebsiteDataStoreConfiguration):
(WebKit::WebsiteDataStoreConfiguration::copy const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):
(TestWebKitAPI::createWebsiteDataStoreAndPrepare):

Canonical link: https://commits.webkit.org/255520@main
